In [1]:
import pandas as pd
from geopy.distance import geodesic
import geopandas as gpd
import lightgbm as lgb

Notes:

ECM = Electronic Control Module (monitors engine sensors + operating conditions)
SPN = Suspect Parameter Number, identifies what system or component is having the issue
FMI = Failure Mode Identifier, explains how the system is failing (too high, too low, short circuit, etc.)

active = fault light (True = ON, False = OFF)
Equpment_Id = truck number

75% derate (SPN = 1569, FMI = 31) reduces engine torque by 25%
idle level derate (SPN = 5246) will require a tow

In [3]:
faults = pd.read_csv("../data/J1939Faults.csv", low_memory=False)

In [5]:
faults['EventTimeStamp'] = pd.to_datetime(faults['EventTimeStamp'])

In [7]:
#MH note: ESS_Id, actionDescription, ecuSoftwareVersion, ecuSerialNumber, ecuModel, ecuMake, ecuSource, faultValue, 
#and MCTNumber are unlikely to provide any predictive value.
faults = faults.drop(columns = ['ESS_Id'
                                , 'actionDescription'
                                , 'ecuSoftwareVersion'
#                                , 'ecuSerialNumber'
                                , 'ecuModel'
                                , 'ecuMake'
                                , 'ecuSource'
                                , 'faultValue'
                                , 'MCTNumber'])

In [9]:
diagnostics = pd.read_csv("../data/VehicleDiagnosticOnboardData.csv")

In [11]:
diagnostics = diagnostics.drop(columns = 'Id')

In [13]:
#pivot wider so each column is a status indicator
diagnostics_pivot = diagnostics.pivot(columns = 'Name', index = 'FaultId', values = 'Value').reset_index()
diagnostics_pivot.columns.name = None

In [ ]:
#check shape to make sure pivoted correctly
print(diagnostics_pivot.shape)
print(diagnostics['FaultId'].nunique())
print(diagnostics['Name'].nunique())

In [15]:
#merge faults and diagnostics
full_df = pd.merge(faults, diagnostics_pivot, how = 'outer', left_on = 'RecordID', right_on = 'FaultId')
full_df = full_df.drop(columns = 'FaultId')

In [77]:
#check shape to make sure nothing hinky in the merge
print(faults.shape)
print(full_df.shape)

(1187335, 12)
(1187335, 37)


In [83]:
diagnostics_pivot.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1187335 entries, 0 to 1187334
Data columns (total 25 columns):
 #   Column                     Non-Null Count    Dtype 
---  ------                     --------------    ----- 
 0   FaultId                    1187335 non-null  int64 
 1   AcceleratorPedal           531889 non-null   object
 2   BarometricPressure         585976 non-null   object
 3   CruiseControlActive        574916 non-null   object
 4   CruiseControlSetSpeed      576458 non-null   object
 5   DistanceLtd                585819 non-null   object
 6   EngineCoolantTemperature   586071 non-null   object
 7   EngineLoad                 585621 non-null   object
 8   EngineOilPressure          586244 non-null   object
 9   EngineOilTemperature       583912 non-null   object
 10  EngineRpm                  586921 non-null   object
 11  EngineTimeLtd              581366 non-null   object
 12  FuelLevel                  502795 non-null   object
 13  FuelLtd                    

In [ ]:
#MH note: There are service locations at (36.0666667, -86.4347222), (35.5883333, -86.4438888), and (36.1950, -83.174722), 
#so you should remove any records in the vicinity of these locations, as fault codes may be tripped when working on the vehicles.

#jeff's code (w/ edits):

service_stations = [
    (36.0666667, -86.4347222),
    (35.5883333, -86.4438888),
    (36.1950, -83.174722)
]

threshold_distance = 0.5  

def is_near_service_station(lat, lon):
    point = (lat, lon)
    for station in service_stations:
        distance = geodesic(point, station).miles
        if distance <= threshold_distance:
            return True
    return False

full_df['IsServiceStation'] = full_df.apply(lambda row: is_near_service_station(row['Latitude'], row['Longitude']), axis=1)

In [17]:
#andrew's code (w/ edits): 
#(lightnight fast, noice!)

# label near service stations

stations = pd.DataFrame(
    {
        "lat": [36.0666667, 35.5883333, 36.1950],
        "lon": [-86.4347222, -86.4438888, -83.174722],
    }
)
threshold_miles = 0.5
threshold_meters = threshold_miles * 1609.34
# create geodataframes with geopandas
gdf_full_df = gpd.GeoDataFrame(
    full_df,
    geometry=gpd.points_from_xy(full_df.Latitude, full_df.Longitude),
    crs="EPSG:4326",  # WGS84 coord ref sys (lat/lon)
)
gdf_stations = gpd.GeoDataFrame(
    stations,
    geometry=gpd.points_from_xy(stations.lat, stations.lon),
    crs="EPSG:4326",
)
target_crs = "EPSG:9311"
# reproject onto new crs for better distance measurement
gdf_full_df_proj = gdf_full_df.to_crs(target_crs)
gdf_stations_proj = gdf_stations.to_crs(target_crs)
# create buffers around stations
station_buf = gdf_stations_proj.geometry.buffer(threshold_meters)
combined_buffer = (
    station_buf.union_all()
)  # turns into single geometry which helps with efficiency
is_within = gdf_full_df_proj.geometry.within(combined_buffer)
full_df["nearStation"] = is_within.values

In [19]:
full_df_notservice = full_df[full_df['nearStation'] == False]

In [ ]:
full_df_notservice.columns

In [91]:
full_df_notservice.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1058051 entries, 0 to 1187334
Data columns (total 37 columns):
 #   Column                     Non-Null Count    Dtype         
---  ------                     --------------    -----         
 0   RecordID                   1058051 non-null  int64         
 1   EventTimeStamp             1058051 non-null  datetime64[ns]
 2   eventDescription           1007093 non-null  object        
 3   ecuSerialNumber            751879 non-null   object        
 4   spn                        1058051 non-null  int64         
 5   fmi                        1058051 non-null  int64         
 6   active                     1058051 non-null  bool          
 7   activeTransitionCount      1058051 non-null  int64         
 8   EquipmentID                1058051 non-null  object        
 9   Latitude                   1058051 non-null  float64       
 10  Longitude                  1058051 non-null  float64       
 11  LocationTimeStamp          1058051 non-nul

In [ ]:
full_df_notservice = full_df_notservice.drop(columns = ['ESS_Id'
                                , 'actionDescription'
                                , 'ecuSoftwareVersion'
#                                , 'ecuSerialNumber'
                                , 'ecuModel'
                                , 'ecuMake'
                                , 'ecuSource'
                                , 'faultValue'
                                , 'MCTNumber'])

In [45]:
derates = full_df_notservice[full_df_notservice['spn'] == 5246]
derates = derates.drop(columns = ['RecordID', 'eventDescription', 'spn', 'LocationTimeStamp', #'Latitude', 'Longitude', 
                                  'AcceleratorPedal', 'CruiseControlActive', 'CruiseControlSetSpeed',
                                  'ServiceDistance', 'CruiseControlActive', 'CruiseControlSetSpeed', 'ParkingBrake', 'nearStation'])
derates = derates[derates['active'] == True]
#derates[derates['fmi'] == True]

In [49]:
pre2019 = derates.loc[(derates['EventTimeStamp'] < '2019-01-01')]

In [69]:
pre2019.groupby('EquipmentID')['fmi'].count().sort_values(ascending=False)#.head(5)

EquipmentID
1524    32
1535    23
1525    15
305     14
1539    14
        ..
1575     1
1571     1
1567     1
1564     1
306      1
Name: fmi, Length: 163, dtype: int64

In [89]:
pre2019[pre2019['EquipmentID'] == '1524']#.head(10)

EventTimeStamp ecuSerialNumber  fmi  active  \
302632 2015-12-10 10:58:47         unknown   16    True   
331209 2016-01-06 14:07:24             NaN   16    True   
331648 2016-01-06 19:45:54             NaN    0    True   
331682 2016-01-06 20:23:45             NaN    0    True   
344278 2016-01-18 15:57:16             NaN   15    True   
344348 2016-01-18 16:57:14             NaN   16    True   
350087 2016-01-22 15:05:25             NaN   16    True   
352286 2016-01-25 13:19:28             NaN   15    True   
353900 2016-01-26 18:21:14             NaN   15    True   
353916 2016-01-26 18:41:52             NaN   16    True   
354853 2016-01-27 12:28:43             NaN   16    True   
361522 2016-02-02 15:09:16             NaN   16    True   
378840 2016-02-16 21:45:56             NaN   16    True   
378866 2016-02-16 22:52:38             NaN    0    True   
400803 2016-03-07 09:10:03             NaN   15    True   
435369 2016-04-11 10:30:55             NaN   15    True   
435452 2016-04-11 11:33:47             NaN   16    True   
441378 2016-04-16 12:56:07             NaN   16    True   
444487 2016-04-19 20:56:49             NaN   16    True   
445207 2016-04-20 12:13:01             NaN   16    True   
445347 2016-04-20 14:11:28             NaN   16    True   
447445 2016-04-22 10:25:42             NaN   16    True   
449785 2016-04-25 13:28:13             NaN   16    True   
449889 2016-04-25 14:56:20             NaN   16    True   
451705 2016-04-27 09:05:50             NaN   16    True   
459710 2016-05-05 16:46:17             NaN   16    True   
459720 2016-05-05 17:07:47             NaN    0    True   
475545 2016-05-24 20:01:58             NaN   16    True   
475650 2016-05-25 01:17:42             NaN    0    True   
475789 2016-05-25 01:39:31             NaN    0    True   
478250 2016-05-27 15:07:38             NaN   16    True   
486368 2016-06-07 12:09:08             NaN    0    True   

        activeTransitionCount EquipmentID   Latitude  Longitude  \
302632                      6        1524  36.067731 -86.429305   
331209                      1        1524  35.873379 -87.859537   
331648                      1        1524  35.031481 -89.891018   
331682                      2        1524  35.031435 -89.891111   
344278                      1        1524  36.121574 -85.777685   
344348                      1        1524  35.986342 -85.013148   
350087                      2        1524  35.270694 -85.889490   
352286                      1        1524  35.557731 -82.548935   
353900                      2        1524  35.992222 -80.484675   
353916                      1        1524  35.870462 -80.707962   
354853                      2        1524  35.955972 -84.958194   
361522                      3        1524  36.167407 -86.529537   
378840                      2        1524  37.013240 -86.287592   
378866                      1        1524  37.096342 -86.055555   
400803                      1        1524  34.012037 -84.567916   
435369                      1        1524  38.221898 -85.476203   
435452                      1        1524  38.175324 -85.693425   
441378                      2        1524  33.724259 -84.913888   
444487                      3        1524  39.645787 -75.924444   
445207                      4        1524  37.358796 -79.973287   
445347                      5        1524  36.724490 -81.913055   
447445                      6        1524  37.112314 -80.442685   
449785                      7        1524  39.846250 -78.273564   
449889                      8        1524  40.114722 -79.393611   
451705                      9        1524  37.022638 -86.249537   
459710                      1        1524  35.633287 -86.246712   
459720                      1        1524  35.625740 -86.244351   
475545                      1        1524  35.466435 -86.056388   
475650                      1        1524  34.116157 -83.984722   
475789                      2        1524  34.098009 -

In [87]:
full_df_notservice.loc[(full_df_notservice['EquipmentID'] == '306') & (full_df_notservice['EventTimeStamp'] > '2018-09-01') & (full_df_notservice['EventTimeStamp'] < '2018-10-10')]

RecordID      EventTimeStamp  \
1014670   1052764 2018-09-04 10:58:10   
1014754   1052848 2018-09-04 13:51:55   
1014759   1052853 2018-09-04 13:57:00   
1015193   1053287 2018-09-06 06:23:04   
1015333   1053427 2018-09-06 10:50:43   
1015526   1053620 2018-09-07 05:45:39   
1015528   1053622 2018-09-07 05:48:43   
1015548   1053642 2018-09-07 06:47:05   
1015659   1053753 2018-09-07 11:22:40   
1021242   1060237 2018-09-25 12:39:12   
1021257   1060252 2018-09-25 12:55:08   
1025771   1064766 2018-10-09 12:26:10   
1025772   1064767 2018-10-09 12:36:01   

                                          eventDescription ecuSerialNumber  \
1014670      Data Drifted High Aftertreatment 1 Outlet NOx        79731935   
1014754  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1014759  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1015193  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1015333  Condition Exists NOx limits exceeded due to In...        79731935   
1015526      Data Drifted High Aftertreatment 1 Outlet NOx        79731935   
1015528  Condition Exists NOx limits exceeded due to In...        79731935   
1015548   Condition Exists Engine Protection Torque Derate        79731935   
1015659                                                NaN        79731935   
1021242  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1021257  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1025771  Incorrect Data J1939 Network #1 Primary Vehicl...      S051400287   
1025772  Incorrect Data J1939 Network #1 Primary Vehicl...             NaN   

          spn  fmi  active  activeTransitionCount EquipmentID   Latitude  ...  \
1014670  3226   20    True                      2         306  35.995879  ...   
1014754   639    2    True                    127         306  36.010601  ...   
1014759   639    2   False                    127         306  36.011990  ...   
1015193   639    2   False                    127         306  36.015462  ...   
1015333  4094   31    True                      1         306  36.038055  ...   
1015526  3226   20    True                      1         306  36.167777  ...   
1015528  4094   31    True                      1         306  36.169907  ...   
1015548  1569   31    True                      1         306  36.173240  ...   
1015659  5246    0    True                      1         306  35.997500  ...   
1021242   639    2    True                    127         306  36.696898  ...   
1021257   639    2   False                    127         306  36.697222  ...   
1025771   639    2    True                    127         306  36.059675  ...   
1025772   639    2   False                    127         306  36.059814  ...   

         IgnStatus IntakeManifoldTemperature LampStatus ParkingBrake  \
1014670       True                     114.8      17407        False   
1014754       True                       140       1279        False   
1014759        NaN                       NaN        255          NaN   
1015193        NaN                       NaN        255          NaN   
1015333       True                       131       2047        False   
1015526       True                       104      17407        False   
1015528       True                     116.6      18431        False   
1015548       True                     145.4      18431         True   
1015659       True                       149      22527        False   
1021242       True                     125.6       1279        False   
1021257        NaN                       NaN        255          NaN   
1025771       True                     138.2       1279        False   
1025772        NaN                       NaN        255          NaN   

        ServiceDistance     Speed SwitchedBatteryVoltage Throttle  \
1014670             NaN  35.00068                    NaN        0   
1014754             NaN  4.417561                    NaN   

useful codes:

sometimes notice early, sometimes right before derate
1569 / 31 = torque derate (limp mode)
5394 / 17 = aftertreatment diesel exhaust system
4094 / 18 = NOx limits exceeded


When evaluating the performance of your model, assume that the cost associated with a missed full derate is approximately $4000 in towing and repairs, and the cost of a false positive prediction is about $500 due to having the truck off the road and serviced unnecessarily. While high accuracy or F1 is nice, we are most interested here in saving the company money, so the final metric to evaulate your model should be the cost savings.